### PR TITLE
[#129082855] authentication thru github API

### PR DESCRIPTION
**Pull request by ferigis**

### CODE REVIEW

File: rebar.config [hunk 1]
```diff
@@ -87,3 +90,13 @@
...
+
+%% == Release ==
+
+{relx, [{release, {spellingci, "0.0.1"},
+         [spellingci]},
+        {dev_mode, true},
+        {include_erts, false},
+        {extended_start_script, true}]}.
+
+        {overrides, [{override, jiffy, [{plugins, [pc]}, {artifacts, ["priv/jiffy.so"]}, {provider_hooks, [{post, [{compile, {pc, compile}}, {clean, {pc, clean}}]}]}]}]}.
```
Comment (elbrujohalcon) [inline rebar.config]: this should not be indented.

File: src/handler/spellingci_callback_handler.erl [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+%trails
+-export([trails/0]).
+
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+%% Trails
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+trails() ->
```
Comment (harenson) [inline src/handler/spellingci_callback_handler.erl]: You're missing the spec here.

File: src/handler/spellingci_callback_handler.erl [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    #{ get => #{ desc => "Github oauth callback"
+               , 'content-type' => "text/plain"}
+     },
+  [trails:trail("/oauth/callback", ?MODULE, [], MsgTrailsMetadata)].
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+%%% Handler Callbacks
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+init(_Type, Req, _Opts) ->
```
Comment (harenson) [inline src/handler/spellingci_callback_handler.erl]: You're missing the spec here.

File: src/handler/spellingci_callback_handler.erl [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+      {ok, Req3, State};
+    {Code, Req2} ->
+      case access_token(Code) of
+        {ok, Token} ->
+          Url = "/",
+          RedirHeaders = [{<<"Location">>, Url}],
+          {ok, Req3} = cowboy_req:reply(302, RedirHeaders, Req2),
+          {ok, Req3, State};
+        {error, Reason} ->
+          _ = lager:log(error, self, "~p", [Reason]),
```
Comment (elbrujohalcon) [inline src/handler/spellingci_callback_handler.erl]: why lager:log instead of lager:error ?

File: src/handler/spellingci_callback_handler.erl [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+  [trails:trail("/oauth/callback", ?MODULE, [], MsgTrailsMetadata)].
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+%%% Handler Callbacks
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+init(_Type, Req, _Opts) ->
+  {ok, Req, #{}}.
+
+handle(Req, State) ->
```
Comment (harenson) [inline src/handler/spellingci_callback_handler.erl]: You're missing the spec here.

File: src/handler/spellingci_callback_handler.erl [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+          {ok, Req3, State};
+        {error, Reason} ->
+          _ = lager:log(error, self, "~p", [Reason]),
+          Body = [<<"Error: ">>, Reason],
+          {ok, Req3} = cowboy_req:reply(400, Headers, Body, Req2),
+          {ok, Req3, State}
+      end
+  end.
+
+terminate(_Reason, _Req, _State) ->
```
Comment (harenson) [inline src/handler/spellingci_callback_handler.erl]: You're missing the spec here.

File: src/handler/spellingci_callback_handler.erl [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+  {ok, ClientSecret} = application:get_env(spellingci, github_client_secret),
+  _ = lager:log(info, self, "~p - ~p", [ClientId, ClientSecret]),
+  Url = "https://github.com/login/oauth/access_token",
+  Headers = [{"Content-Type", "application/x-www-form-urlencoded"},
+             {"Accept", "application/json"}],
+  Body = ["code=", Code,
+          "&client_id=", ClientId,
+          "&client_secret=", ClientSecret],
+  Opts = [{ssl_options, [{depth, 0}]}],
+  case ibrowse:send_req(Url, Headers, post, Body, Opts) of
```
Comment (elbrujohalcon) [inline src/handler/spellingci_callback_handler.erl]: Don't use `ibrowse`. Use `hackney`.

File: src/handler/spellingci_login_handler.erl [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+  ++ "client_id=" ++ ClientId
+  ++ "&scope=" ++ Scope,
+  Headers = [{<<"Location">>, Url}],
+  Body = [],
+  {ok, Req2} = cowboy_req:reply(302, Headers, Body, Req),
+  {ok, Req2, State}.
+
+-spec terminate(term(), cowboy_req:req(), map()) -> ok.
+terminate(_Reason, _Req, _State) ->
+    ok.
```
Comment (elbrujohalcon) [inline src/handler/spellingci_login_handler.erl]: bad indentation

File: src/spellingci.app.src [hunk 9]
```diff
@@ -7,7 +7,11 @@
     kernel,
     stdlib,
     mnesia,
-    sumo_rest
+    sumo_rest,
+    ibrowse,
```
Comment (elbrujohalcon) [inline src/spellingci.app.src]: don't use ibrowse, use hackney

File: src/handler/spellingci_callback_handler.erl [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+    {ok, "200", _RespHeaders, RespBody} ->
+      JsonBody = jiffy:decode(RespBody, [return_maps]),
+      case maps:is_key(<<"access_token">>, JsonBody) of
```
Comment (harenson) [inline src/handler/spellingci_callback_handler.erl]: You could also write this case like this:

```erlang
      case maps:get(<<"access_token">>, JsonBody, undefined) of
        undefined ->
          {error, RespBody};
        Token ->
          {ok, Token}
      end;
```

File: src/handler/spellingci_login_handler.erl [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+        ]).
+
+%trails
+-export([trails/0]).
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+%% Trails
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+trails() ->
```
Comment (harenson) [inline src/handler/spellingci_login_handler.erl]: Missing spec.

File: src/handler/spellingci_login_handler.erl [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+               , 'content-type' => "text/plain"}
+     },
+  [trails:trail("/oauth/login", ?MODULE, [], MsgTrailsMetadata)].
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+%% Cowboy callbacks
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+%% @hidden
+init(_Transport, Req, []) ->
```
Comment (harenson) [inline src/handler/spellingci_login_handler.erl]: Missing spec.

File: src/handler/spellingci_login_handler.erl [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+  {ok, Req, #{}}.
+
+%% @hidden
+handle(Req, State) ->
```
Comment (harenson) [inline src/handler/spellingci_login_handler.erl]: Missing spec.

File: src/handler/spellingci_login_handler.erl [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+-export([trails/0]).
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+%% Trails
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+trails() ->
+  MsgTrailsMetadata =
+    #{ get => #{ desc => "Starts the oauth login process with Github"
+               , 'content-type' => "text/plain"}
```
Comment (harenson) [inline src/handler/spellingci_login_handler.erl]: Move the closing curly brace (`{`) to the next line and align it with the comma.

**Pull request merged**